File: dental_app_mobile/lib/core/socket_service.dart
import 'dart:async';

import 'package:socket_io_client/socket_io_client.dart' as io;

import 'config.dart';

/// Gerçek zamanlı bildirim payload'ı (backend 'notification' eventi).
class LiveNotification {
  final String type;
  final String title;
  final String message;
  final Map<String, dynamic>? data;
  final String? timestamp;

  LiveNotification({
    required this.type,
    required this.title,
    required this.message,
    this.data,
    this.timestamp,
  });

  factory LiveNotification.fromJson(Map<String, dynamic> json) {
    return LiveNotification(
      type: json['type']?.toString() ?? '',
      title: json['title']?.toString() ?? '',
      message: json['message']?.toString() ?? '',
      data: json['data'] is Map<String, dynamic>
          ? json['data'] as Map<String, dynamic>
          : null,
      timestamp: json['timestamp']?.toString(),
    );
  }
}

/// Socket.IO bildirim istemcisi.
///
/// Backend sözleşmesi (src/services/notificationHub.js):
///  - Kimlik doğrulama: handshake auth.token içinde JWT
///  - Eventler: 'connected', 'notification'
class SocketService {
  io.Socket? _socket;
  final _controller = StreamController<LiveNotification>.broadcast();

  Stream<LiveNotification> get notifications => _controller.stream;

  void connect(String accessToken) {
    disconnect();

    final socket = io.io(
      AppConfig.socketUrl,
      io.OptionBuilder()
          .setTransports(['websocket'])
          .setAuth({'token': accessToken})
          .setPath('/socket.io/')
          .enableReconnection()
          .setReconnectionDelay(2000)
          .build(),
    );

    socket.on('notification', (data) {
      if (data is Map) {
        _controller.add(
          LiveNotification.fromJson(Map<String, dynamic>.from(data)),
        );
      }
    });

    socket.connect();
    _socket = socket;
  }

  void disconnect() {
    _socket?.dispose();
    _socket = null;
  }

  void dispose() {
    disconnect();
    _controller.close();
  }
}

File: dental-app-web/src/app/core/services/socket.service.ts
import { Injectable } from '@angular/core';
import { io, Socket } from 'socket.io-client';
import { Subject } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Notification } from '../models/models';

/**
 * Socket.IO tabanlı gerçek zamanlı bildirim servisi.
 * Backend sözleşmesi (src/services/notificationHub.js):
 *  - Kimlik doğrulama: handshake auth.token içinde JWT
 *  - Eventler: 'connected' (bağlantı onayı), 'notification' ({type,title,message,data,timestamp})
 *  - İstemci -> sunucu: 'subscribe'/'unsubscribe' ({channel})
 */
@Injectable({
    providedIn: 'root'
})
export class SocketService {
    private socket?: Socket;
    private notificationSubject = new Subject<Notification>();
    public notification$ = this.notificationSubject.asObservable();

    connect(accessToken: string): void {
        if (this.socket?.connected) {
            return;
        }
        this.disconnect();

        this.socket = io(environment.socketUrl, {
            path: '/socket.io/',
            auth: { token: accessToken },
            transports: ['websocket', 'polling'],
            reconnection: true,
            reconnectionAttempts: Infinity,
            reconnectionDelay: 2000
        });

        this.socket.on('notification', (notification: Notification) => {
            this.notificationSubject.next(notification);
        });

        this.socket.on('connected', (data: unknown) => {
            console.log('Bildirim sunucusuna bağlanıldı:', data);
        });

        this.socket.on('connect_error', (err: Error) => {
            console.warn('Bildirim bağlantı hatası:', err.message);
        });
    }

    disconnect(): void {
        if (this.socket) {
            this.socket.disconnect();
            this.socket = undefined;
        }
    }

    subscribe(channel: string): void {
        this.socket?.emit('subscribe', { channel });
    }

    unsubscribe(channel: string): void {
        this.socket?.emit('unsubscribe', { channel });
    }
}
